File: market/margin.go
package market

// PositionSide mirrors Side but reads more naturally when describing
// a held position rather than an order direction.
type PositionSide int

const (
	Long PositionSide = iota
	Short
)

// Position represents one agent's holding in one instrument.
type Position struct {
	Side      PositionSide
	Quantity  float64
	EntryCost float64 // total cash paid/received at entry (before leverage math)
}

// Account tracks one agent's capital and leverage state. This is
// deliberately separate from "Agent" (the decision-making layer) — an
// Account is just the ledger: cash, positions, margin usage. Any agent
// type (retail, hedge fund, market maker) holds one of these.
type Account struct {
	AgentID string
	Cash    float64

	// Positions keyed by instrument symbol.
	Positions map[string]*Position

	// MaxLeverage is the max (position value / equity) this account is
	// allowed to run. 1.0 = no leverage (fully cash-backed). A retail
	// account might be capped at 2x, a hedge fund at 5-10x, etc.
	MaxLeverage float64

	// MaintenanceMarginRatio is the equity/position-value floor before
	// a margin call is triggered. Typically lower than 1/MaxLeverage —
	// e.g. you might be allowed to *open* at 5x leverage but only get
	// margin-called once equity/position-value drops below 10%.
	MaintenanceMarginRatio float64
}

func NewAccount(agentID string, startingCash float64, maxLeverage, maintenanceMarginRatio float64) *Account {
	return &Account{
		AgentID:                agentID,
		Cash:                   startingCash,
		Positions:              make(map[string]*Position),
		MaxLeverage:            maxLeverage,
		MaintenanceMarginRatio: maintenanceMarginRatio,
	}
}

// PositionValue returns the current mark-to-market value of a position
// at the given current price (always positive — direction is tracked
// separately via Side, this is magnitude of exposure).
func (p *Position) PositionValue(currentPrice float64) float64 {
	return p.Quantity * currentPrice
}

// UnrealizedPnL computes floating profit/loss on a position at the
// given current price.
func (p *Position) UnrealizedPnL(currentPrice float64) float64 {
	marketValue := p.Quantity * currentPrice
	if p.Side == Long {
		return marketValue - p.EntryCost
	}
	// Short: profit when price falls below entry.
	return p.EntryCost - marketValue
}

// Equity computes total account equity (cash + unrealized P&L across
// all held positions) given a map of current prices by symbol.
func (a *Account) Equity(currentPrices map[string]float64) float64 {
	equity := a.Cash
	for symbol, pos := range a.Positions {
		if price, ok := currentPrices[symbol]; ok {
			equity += pos.UnrealizedPnL(price)
		}
	}
	return equity
}

// TotalPositionValue sums the mark-to-market magnitude of all exposure,
// used as the denominator for margin ratio checks.
func (a *Account) TotalPositionValue(currentPrices map[string]float64) float64 {
	var total float64
	for symbol, pos := range a.Positions {
		if price, ok := currentPrices[symbol]; ok {
			total += pos.PositionValue(price)
		}
	}
	return total
}

// MarginRatio returns equity / total position value. A ratio of 1.0
// means no leverage in use; lower ratios mean more leverage/risk.
// Returns (0, false) if the account holds no positions (ratio undefined).
func (a *Account) MarginRatio(currentPrices map[string]float64) (float64, bool) {
	posValue := a.TotalPositionValue(currentPrices)
	if posValue <= 0 {
		return 0, false
	}
	return a.Equity(currentPrices) / posValue, true
}

// IsUnderMaintenanceMargin reports whether this account has fallen
// below its maintenance margin threshold and should be liquidated.
func (a *Account) IsUnderMaintenanceMargin(currentPrices map[string]float64) bool {
	ratio, hasPositions := a.MarginRatio(currentPrices)
	if !hasPositions {
		return false
	}
	return ratio < a.MaintenanceMarginRatio
}

// AvailableBuyingPower returns how much *additional* position value this
// account could still take on given its current equity and max leverage.
func (a *Account) AvailableBuyingPower(currentPrices map[string]float64) float64 {
	equity := a.Equity(currentPrices)
	maxExposure := equity * a.MaxLeverage
	used := a.TotalPositionValue(currentPrices)
	remaining := maxExposure - used
	if remaining < 0 {
		return 0
	}
	return remaining
}

// ApplySettledFill updates acct's position in symbol to reflect one
// settled fill: side is the direction *from this account's
// perspective* (Buy fills add/open a Long, Sell fills add/open a
// Short), quantity and price describe the fill itself. This is the
// one canonical settlement path — the sim loop's tick logic should
// call this for every fill an account is a party to, rather than each
// call site reimplementing position bookkeeping independently.
//
// Handles three cases: no existing position (open a new one),
// same-direction fill (add to the position, weighted into
// EntryCost), and opposing-direction fill (reduce the existing
// position, or fully close and flip to the other side if the fill
// size exceeds what was held).
func ApplySettledFill(acct *Account, symbol string, side Side, quantity, price float64) {
	positionSide := Long
	if side == Sell {
		positionSide = Short
	}

	pos, ok := acct.Positions[symbol]
	if !ok {
		acct.Positions[symbol] = &Position{Side: positionSide, Quantity: quantity, EntryCost: quantity * price}
		return
	}

	if pos.Side == positionSide {
		pos.Quantity += quantity
		pos.EntryCost += quantity * price
		return
	}

	// Opposing fill: reduces the existing position, or fully closes
	// and flips to the other side if this fill's quantity exceeds
	// what was held.
	if quantity < pos.Quantity {
		pos.Quantity -= quantity
		return
	}
	remainder := quantity - pos.Quantity
	pos.Side = positionSide
	pos.Quantity = remainder
	pos.EntryCost = remainder * price
}

// LiquidationEngine scans a set of accounts each tick and generates
// forced-close market orders for any account under maintenance margin.
// This is intentionally decoupled from the OrderBook: it only produces
// orders. The caller (the sim's main tick loop) is responsible for
// submitting those orders to the book, which is what lets a forced
// liquidation itself move price and potentially trigger further
// liquidations elsewhere — the cascade emerges from that loop, not
// from anything hardcoded here.
type LiquidationEngine struct{}

// ForcedOrder pairs a generated liquidation order with the symbol it
// applies to and the account it was generated from, so the caller can
// route it to the right book and update the right account afterward.
type ForcedOrder struct {
	Symbol  string
	Order   Order
	Account *Account
}

// ScanForLiquidations checks every account against current prices and
// returns market orders that fully close out any position(s) held by
// accounts under maintenance margin. currentPrices must cover every
// symbol any account might hold.
func (LiquidationEngine) ScanForLiquidations(accounts []*Account, currentPrices map[string]float64) []ForcedOrder {
	var forced []ForcedOrder

	for _, acct := range accounts {
		if !acct.IsUnderMaintenanceMargin(currentPrices) {
			continue
		}
		for symbol, pos := range acct.Positions {
			if pos.Quantity <= 0 {
				continue
			}
			// Closing a long = sell; closing a short = buy back.
			side := Sell
			if pos.Side == Short {
				side = Buy
			}
			forced = append(forced, ForcedOrder{
				Symbol: symbol,
				Order: Order{
					AgentID:  acct.AgentID,
					Side:     side,
					Quantity: pos.Quantity,
					IsMarket: true,
				},
				Account: acct,
			})
		}
	}

	return forced
}

File: src/market/margin.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use super::orderbook::{Order, Side};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum PositionSide {
    Long,
    Short,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub side: PositionSide,
    pub quantity: f64,
    pub entry_cost: f64, // total cash paid/received at entry
    pub opened_at: i64,   // Unix timestamp nanos/millis
}

impl Position {
    pub fn new(side: PositionSide, quantity: f64, entry_cost: f64) -> Self {
        Self {
            side,
            quantity,
            entry_cost,
            opened_at: chrono::Utc::now().timestamp_millis(),
        }
    }

    pub fn position_value(&self, current_price: f64) -> f64 {
        self.quantity * current_price
    }

    pub fn unrealized_pnl(&self, current_price: f64) -> f64 {
        let market_value = self.quantity * current_price;
        match self.side {
            PositionSide::Long => market_value - self.entry_cost,
            PositionSide::Short => self.entry_cost - market_value,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    pub agent_id: String,
    pub cash: f64,
    pub positions: HashMap<String, Position>,
    pub max_leverage: f64,
    pub maintenance_margin_ratio: f64,
}

impl Account {
    pub fn new(
        agent_id: impl Into<String>,
        starting_cash: f64,
        max_leverage: f64,
        maintenance_margin_ratio: f64,
    ) -> Self {
        Self {
            agent_id: agent_id.into(),
            cash: starting_cash,
            positions: HashMap::new(),
            max_leverage,
            maintenance_margin_ratio,
        }
    }

    pub fn equity(&self, current_prices: &HashMap<String, f64>) -> f64 {
        let mut equity = self.cash;
        for (symbol, pos) in &self.positions {
            if let Some(&price) = current_prices.get(symbol) {
                equity += pos.unrealized_pnl(price);
            }
        }
        equity
    }

    pub fn total_position_value(&self, current_prices: &HashMap<String, f64>) -> f64 {
        let mut total = 0.0;
        for (symbol, pos) in &self.positions {
            if let Some(&price) = current_prices.get(symbol) {
                total += pos.position_value(price);
            }
        }
        total
    }

    pub fn margin_ratio(&self, current_prices: &HashMap<String, f64>) -> Option<f64> {
        let pos_val = self.total_position_value(current_prices);
        if pos_val <= 0.0 {
            None
        } else {
            Some(self.equity(current_prices) / pos_val)
        }
    }

    pub fn is_under_maintenance_margin(&self, current_prices: &HashMap<String, f64>) -> bool {
        match self.margin_ratio(current_prices) {
            Some(ratio) => ratio < self.maintenance_margin_ratio,
            None => false,
        }
    }

    pub fn available_buying_power(&self, current_prices: &HashMap<String, f64>) -> f64 {
        let eq = self.equity(current_prices);
        let max_exposure = eq * self.max_leverage;
        let used = self.total_position_value(current_prices);
        (max_exposure - used).max(0.0)
    }
}

pub fn apply_settled_fill(acct: &mut Account, symbol: &str, side: Side, quantity: f64, price: f64) {
    let position_side = match side {
        Side::Buy => PositionSide::Long,
        Side::Sell => PositionSide::Short,
    };

    let pos = match acct.positions.get_mut(symbol) {
        None => {
            acct.positions.insert(
                symbol.to_string(),
                Position {
                    side: position_side,
                    quantity,
                    entry_cost: quantity * price,
                    opened_at: chrono::Utc::now().timestamp_millis(),
                },
            );
            return;
        }
        Some(p) => p,
    };

    if pos.side == position_side {
        pos.quantity += quantity;
        pos.entry_cost += quantity * price;
        return;
    }

    // Opposing fill: reduces existing position, or fully closes and flips
    let avg_cost = pos.entry_cost / pos.quantity;
    let closed_qty = quantity.min(pos.quantity);

    let realized_pnl = match pos.side {
        PositionSide::Long => closed_qty * (price - avg_cost),
        PositionSide::Short => closed_qty * (avg_cost - price),
    };
    acct.cash += realized_pnl;

    if quantity < pos.quantity {
        pos.quantity -= quantity;
        pos.entry_cost = pos.quantity * avg_cost;
        return;
    }

    let remainder = quantity - pos.quantity;
    if remainder <= 1e-9 {
        acct.positions.remove(symbol);
    } else {
        let pos = acct.positions.get_mut(symbol).unwrap();
        pos.side = position_side;
        pos.quantity = remainder;
        pos.entry_cost = remainder * price;
        pos.opened_at = chrono::Utc::now().timestamp_millis();
    }
}

#[derive(Debug, Clone)]
pub struct ForcedOrder {
    pub symbol: String,
    pub order: Order,
    pub agent_id: String,
}

#[derive(Debug, Default, Clone, Copy)]
pub struct LiquidationEngine;

impl LiquidationEngine {
    pub fn scan_for_liquidations(
        accounts: &[&Account],
        current_prices: &HashMap<String, f64>,
    ) -> Vec<ForcedOrder> {
        let mut forced = Vec::new();
        for acct in accounts {
            if !acct.is_under_maintenance_margin(current_prices) {
                continue;
            }
            for (symbol, pos) in &acct.positions {
                if pos.quantity <= 1e-9 {
                    continue;
                }
                let side = match pos.side {
                    PositionSide::Long => Side::Sell,
                    PositionSide::Short => Side::Buy,
                };
                forced.push(ForcedOrder {
                    symbol: symbol.clone(),
                    order: Order {
                        id: 0,
                        agent_id: acct.agent_id.clone(),
                        side,
                        price: 0.0,
                        quantity: pos.quantity,
                        is_market: true,
                    },
                    agent_id: acct.agent_id.clone(),
                });
            }
        }
        forced
    }
}
